readthedocs/projects/tasks/builds.py

```diff
@@ -0,0 +1,1020 @@
+                del self.__dict__[attribute]
+
+    def execute(self):
+        environment = self.environment_class(
+            project=self.project,
+            version=self.version,
+            environment=self.get_vcs_env_vars(),
+        )
+
+        with environment:
+            before_vcs.send(
+                sender=self.version,
+                environment=environment,
+            )
+            self.update_versions_from_repository(environment)
+
+    def update_versions_from_repository(self, environment):
+        """
+        Update Read the Docs versions from VCS repository.
+
+        Depending if the VCS backend supports remote listing, we just list its branches/tags
+        remotely or we do a full clone and local listing of branches/tags.
+        """
+        version_repo = self.get_vcs_repo(environment)
+        if any([
+                not version_repo.supports_lsremote,
+                not self.project.has_feature(Feature.VCS_REMOTE_LISTING),
+        ]):
+            log.info('Syncing repository via full clone.')
+            self.sync_repo(environment)
+        else:
+            log.info('Syncing repository via remote listing.')
+            self.sync_versions(version_repo)
+
+
+@app.task(
+    base=SyncRepositoryTask,
+    bind=True,
+)
+def sync_repository_task(self, version_id=None):
+    # NOTE: `before_start` is new on Celery 5.2.x, but we are using 5.1.x currently.
+    self.before_start(self.request.id, self.request.args, self.request.kwargs)
+
+    self.execute()
+
+
+class UpdateDocsTask(SyncRepositoryMixin, Task):
+
+    """
+    The main entry point for updating documentation.
+
+    It handles all of the logic around whether a project is imported, was
+    created or a webhook is received. Then it will sync the repository and
+    build all the documentation formats and upload them to the storage.
+    """
+
+    name = __name__ + '.update_docs_task'
+    autoretry_for = (
+        BuildMaxConcurrencyError,
+    )
+    max_retries = 5  # 5 per normal builds, 25 per concurrency limited
+    default_retry_delay = 7 * 60
+
+    # Expected exceptions that will be logged as info only and not retried
+    throws = (
+        DuplicatedBuildError,
+        VersionLockedError,
+        ProjectBuildsSkippedError,
+        ConfigError,
+        YAMLParseError,
+    )
+
+    acks_late = True
+    track_started = True
+
+    # These values have to be dynamic based on project
+    time_limit = None
+    soft_time_limit = None
+
+    Request = BuildRequest
+
+    # TODO:
+    # 1. raise Reject(requeue=False) on duplicated builds
+    # 2. use a global `task_cls` (https://docs.celeryproject.org/en/latest/userguide/tasks.html#app-wide-usage) to logs actions
+    # 3. use CELERY_IMPORTS to register the tasks (https://docs.celeryproject.org/en/latest/userguide/configuration.html#std-setting-imports)
+    # 4. use CELERY_TASK_IGNORE_RESULT=True since we are not using the result at all
```

Comment:
I'm not taking them in this PR, no. These are notes that I found useful from the documentation while reading it and that could apply to our use cases. I'm not sure they are _good_, tho. We can come back to them later and check if it's worth the effort.

Reply:
It might be worth noting this somewhere, perhaps just start an issue preemptively? We'll have some issues come out of the left over TODOs here at the end I'm sure, but don't need to track every little change.

Reply:
Yeah. I plan to come back to the notes linked in the description and open the corresponding issues.